tests/interfaces/default_qubit_2_integration/test_autograd_qnode.py

```diff
@@ -0,0 +1,2030 @@
+            grad = qml.grad(cost)(a, b)
+
+        assert grad == tuple()
+
+    def test_matrix_parameter(self, interface, dev, diff_method, grad_on_execution, tol):
+        """Test that the autograd interface works correctly
+        with a matrix parameter"""
+        U = np.array([[0, 1], [1, 0]], requires_grad=False)
+        a = np.array(0.1, requires_grad=True)
+
+        @qnode(
+            dev, diff_method=diff_method, interface=interface, grad_on_execution=grad_on_execution
+        )
+        def circuit(U, a):
+            qml.QubitUnitary(U, wires=0)
+            qml.RY(a, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        res = circuit(U, a)
+
+        if diff_method == "finite-diff":
+            assert circuit.qtape.trainable_params == [1]
+
+        res = qml.grad(circuit)(U, a)
+        assert np.allclose(res, np.sin(a), atol=tol, rtol=0)
+
+    def test_gradient_non_differentiable_exception(
+        self, interface, dev, diff_method, grad_on_execution
+    ):
+        """Test that an exception is raised if non-differentiable data is
+        differentiated"""
+
+        @qnode(dev, interface=interface, diff_method=diff_method)
+        def circuit(data1):
+            qml.templates.AmplitudeEmbedding(data1, wires=[0, 1])
+            return qml.expval(qml.PauliZ(0))
+
+        grad_fn = qml.grad(circuit, argnum=0)
+        data1 = np.array([0, 1, 1, 0], requires_grad=False) / np.sqrt(2)
+
+        with pytest.raises(qml.numpy.NonDifferentiableError, match="is non-differentiable"):
+            grad_fn(data1)
+
+    def test_differentiable_expand(self, interface, dev, diff_method, grad_on_execution, tol):
+        """Test that operation and nested tape expansion
+        is differentiable"""
+
+        if diff_method == "spsa":
+            np.random.seed(SEED_FOR_SPSA)
+            tol = TOL_FOR_SPSA
+
+        class U3(qml.U3):
+            def expand(self):
+                theta, phi, lam = self.data
+                wires = self.wires
+
+                with qml.queuing.AnnotatedQueue() as q_tape:
+                    qml.Rot(lam, theta, -lam, wires=wires)
+                    qml.PhaseShift(phi + lam, wires=wires)
+
+                tape = QuantumScript.from_queue(q_tape)
+                return tape
+
+        a = np.array(0.1, requires_grad=False)
+        p = np.array([0.1, 0.2, 0.3], requires_grad=True)
+
+        @qnode(
+            dev, diff_method=diff_method, interface=interface, grad_on_execution=grad_on_execution
+        )
+        def circuit(a, p):
+            qml.RX(a, wires=0)
+            U3(p[0], p[1], p[2], wires=0)
+            return qml.expval(qml.PauliX(0))
+
+        res = circuit(a, p)
+        expected = np.cos(a) * np.cos(p[1]) * np.sin(p[0]) + np.sin(a) * (
+            np.cos(p[2]) * np.sin(p[1]) + np.cos(p[0]) * np.cos(p[1]) * np.sin(p[2])
+        )
+        # assert isinstance(res, np.ndarray)
+        # assert res.shape == ()
```

Comment:
```suggestion
        assert isinstance(res, (np.float64, np.tensor))
```

Reply:
The problem is that it becomes float instead. It's never re-cast into the starting interface.  Something we can fix, but I don't think that needs to block this PR.

Reply:
Going to resolve the comments of this type. We can come back in and make sure the output of sampling is always coverted to the correct tensor.